### ISSUE TITLE
Mod RA has so much lag it's unplayable

### ISSUE THREAD

**Issue by SoMuchLag** [OpenRA/OpenRA#8755]
Play mod RA on any large map and eventually the game freezes for 5 seconds every 3 seconds or so.
With many debug messages similar to: AI (X) has nowhere to place apwr.

The freeze is so bad it is simply unplayable.

How about stopping the AI from trying the same thing over and over again every 5 seconds?
If something failed once, try again in 10 seconds, failed again, try again in 20 seconds, failed again, try again in 40 seconds, etc.

Btw have you guys tried using LuaJIT?

**Comment by GraionDilach** [comment]
Sounds like the AI ran out of base space around the MCV. This has nothing to do with Lua.

In TS/RA2, there was a property for the AI, that if it cannot place the building in x amount of tries it abandons the production of that one. That should work a ton better compared to your idea, because yours have the danger of deadlocking the AI production queue for unnecessarily long times.

**Comment by Unit158** [comment]
It is not feasable to inteface LuaJIT with our C# code, and it would be extremely dumb to do so even if we could for security reasons.

**Comment by SoMuchLag** [comment]
> GraionDilach:
>
> Sounds like the AI ran out of base space around the MCV. This has nothing to do with Lua.
> In TS/RA2, there was a property for the AI, that if it cannot place the building in x amount of tries it abandons the production of that one. That should work a ton better compared to your idea, because yours have the danger of deadlocking the AI production queue for unnecessarily long times.

Ok let's get picky then.

This problem was reported a long time ago, since it's not fixed yet I am assuming this problem is requiring too much effort to fix for the devs. So I offered a quick and dirty idea for a fix.

The key here is RETRY, your idea has the danger of AI abandoning production of everything just because vehicles/people happen to be nearby the base at the wrong time, multiple times.

The production queue is already "deadlocking", it's already rebuilding over and over again, my quick and dirty fix simply tell it to increase the delay between them.

> Unit158 
>
> It is not feasable to inteface LuaJIT with our C# code, and it would be extremely dumb to do so even if we could for security reasons.

In other words, you suck at programming and security.

Tell me genius, how smart is it to create a loop that freeze the whole game every few seconds by building the same crap over and over again in an area with little space?

**Comment by GraionDilach** [comment]
Oh look, we've got a badass over here!

You have better ideas? Doesn't seem so. Have you seen the TS/RA2 AI failing with buildings many times? Fun fact, WW set this to two, so the AI gives up after two tries... this worked pretty well even on yaps where the mapper didn't added water, but forgot to disable the shipyards...

My suggestion works on a game which never even had such issues.

But if we suck at coding and security, please, fix it yourself, teach us, you grandmaster, you must know it a lot better than us!

**Comment by pchote** [comment]
Please keep the discussion civil.  Being rude to the people who might otherwise be willing to help you out is an *extremely* bad way to get things done.

Our development team is small and made up entirely of people volunteering their scant free time.  There are currently no active developers who are able to work on the AI, so these issues cannot be prioritized.

**Comment by Mailaender** [comment]
> Play mod RA on any large map and eventually the game freezes for 5 seconds every 3 seconds or so. With many debug messages similar to: AI (X) has nowhere to place apwr.

This is a duplicate of https://github.com/OpenRA/OpenRA/issues/4717. Closing.

> Btw have you guys tried using LuaJIT?

The [Lua API](https://github.com/OpenRA/OpenRA/wiki/Lua-API) is only used for mission scripting, not AI although that is also an idea. https://github.com/OpenRA/OpenRA/issues/4947
